genwinsyscallproto/genwinsyscalldef.go

```diff
@@ -0,0 +1,554 @@
+	default:
+		if p.Kind == flags.ElementType_PTR {
+			b.WriteString("*")
+		}
+		if visited == nil {
+			visited = make(map[*winmd.SigType]struct{})
+		}
+		visited[p] = struct{}{}
+		return c.writeTypeValue(b, p.Value, visited)
+	}
+	return nil
+}
+
+func (c *Context) writeTypeValue(b io.StringWriter, value any, visited map[*winmd.SigType]struct{}) error {
+	switch v := value.(type) {
+	case winmd.CodedIndex:
+		switch v.Tag {
+		case coded.TypeDefOrRefOrSpec_TypeDef:
+			record, err := c.Metadata.Tables.TypeDef.Record(v.Index)
+			if err != nil {
+				return err
+			}
+			writeEscapedUpper(b, record.Name.String())
+
+		case coded.TypeDefOrRefOrSpec_TypeRef:
+			record, err := c.Metadata.Tables.TypeRef.Record(v.Index)
+			if err != nil {
+				return err
+			}
+			c.UsedTypeRefs[v.Index] = struct{}{}
+			writeEscapedUpper(b, record.Name.String())
+
+		default:
+			return fmt.Errorf("unexpected coded index value: %#v", v)
+		}
+
+	// Types can nest. A pointer to another type is a very common case.
+	case winmd.SigType:
+		return c.writeTypeCore(b, &v, visited)
+	case winmd.SigArray:
+		b.WriteString("[]")
+		return c.writeTypeCore(b, &v.Type, visited)
+
+	default:
+		return fmt.Errorf("unexpected type value: %#v", value)
+	}
+	return nil
+}
+
+func (c *Context) WriteTypeDef(b io.StringWriter, i winmd.Index) error {
+	def, err := c.Metadata.Tables.TypeDef.Record(i)
+	if err != nil {
+		return err
+	}
+	if def.Flags&flags.TypeAttributes_ClassSemanticsMask == flags.TypeAttributes_Interface {
+		// TODO: Handle interfaces. Currently writes a struct with no members.
+		return c.writeTypeDefStruct(b, def)
+	}
+	switch def.Extends.Tag {
+	case coded.TypeDefOrRef_TypeRef:
+		// TODO: Keep track of this index rather than looking it up for each enum type.
+		record, err := c.Metadata.Tables.TypeRef.Record(def.Extends.Index)
+		if err != nil {
+			return err
+		}
+		if record.Namespace.String() == "System" && record.Name.String() == "Enum" {
+			return c.writeTypeDefEnum(b, def)
+		}
+		if _, ok := c.TypeDefNativeTypedefAttribute[i]; ok {
+			return c.writeTypeDefNative(b, def)
+		}
+		return c.writeTypeDefStruct(b, def)
+	default:
+		return fmt.Errorf("unexpected type extends coded index: %#v", def.Extends)
+	}
+}
+
+func (c *Context) writeTypeDefEnum(b io.StringWriter, def *winmd.TypeDef) error {
+	b.WriteString("type ")
+	writeEscapedUpper(b, def.Name.String())
+
+	// Per §I.8.5.2 CLS Rule 7, the underlying type is the type of the field "__value". Find it.
+	var underlyingType *winmd.SigType
+
+	type member struct {
+		Name     string
+		HexValue string
+	}
+	// The number of enum members is the total number of fields minus the special "value__".
+	members := make([]member, 0, def.FieldList.End-def.FieldList.Start-1)
+
+	for i := def.FieldList.Start; i < def.FieldList.End; i++ {
+		fd, err := c.Metadata.Tables.Field.Record(i)
+		if err != nil {
+			return err
+		}
+		signature, err := c.Metadata.FieldSignature(fd.Signature)
```

Comment:
`signature` is only used if `fd.Name.String() == "value__"`, so there is no need to always decode it.